File: 345reverse-vowels-of-a-string/twoPointer.go
package main

// 双指针. 找到元音字母则替换

// 时间 O(n), 空间 O(1)

import (
	"fmt"
)

func main() {
	s := "leetcode"
	fmt.Println(reverseVowels(s))

	s = "hello"
	fmt.Println(reverseVowels(s))
}

func reverseVowels(s string) string {

	vowels := map[byte]byte{
		'a': 'a', 'e': 'e', 'i': 'i', 'o': 'o', 'u': 'u',
		'A': 'A', 'E': 'E', 'I': 'I', 'O': 'O', 'U': 'U',
	}
	res := []byte(s)

	p1 := 0
	p2 := len(s) - 1

	for p1 < p2 {
		v1, ok1 := vowels[s[p1]]
		v2, ok2 := vowels[s[p2]]

		if ok1 && !ok2 {
			p2--
			continue
		}

		if !ok1 && ok2 {
			p1++
			continue
		}

		if ok1 && ok2 {
			res[p1], res[p2] = v2, v1
		}
		p1++
		p2--
	}

	return string(res)
}

File: 345reverse-vowels-of-a-string/twoPointer.c
#include <stdio.h>
#include <string.h>
bool isVowels(char c);
char * reverseVowels(char * s);


int main() {
    char s[] ="leetcode";
    printf("%s\n", reverseVowels(s));
    char s2[] = "hello";
    printf("%s\n", reverseVowels(s2));
}


char * reverseVowels(char * s){
    int start = 0;
    int end = strlen(s) - 1;

    while(start < end) {
        while(!isVowels(s[start]) && start < end) start++;
        while(!isVowels(s[end]) && start < end) end--;
        char tmp = s[start];
        s[start] = s[end];
        s[end] = tmp;

        start++;
        end--;
    }

    return s;
}

bool isVowels(char c) {
    if (c == 'a' || c =='e'|| c == 'i'|| c == 'o'|| c == 'u'|| 
        c == 'A'|| c == 'E'|| c == 'I'|| c == 'O'|| c == 'U' 
    ) {
        return true;
    }
    return false;
}